### PR TITLE
Allocator rework

### PR DESCRIPTION
**Pull request by bretambrose**
Add stateful reference to underlying c-common allocator.

An internal #define to make the default constructor invisible verifies that crt logic always manually selects an aws-c-common allocator.

There are some potentially questionable decision made here.

Also, certain construction paths (in particular, Vector<String> (count)) required a code change because the internal implementation was not cloning the Vector's allocator when in-place building the reserved Strings.  This makes me nervous.

This will require a new release

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/iot/MqttClient.h [hunk 1]
```diff
@@ -29,7 +29,7 @@ namespace Aws
         class AWS_CRT_CPP_API MqttClientConnectionConfig final
         {
           public:
-            MqttClientConnectionConfig() noexcept;
+            // MqttClientConnectionConfig() noexcept;
```
Comment (justinboswell) [inline include/aws/iot/MqttClient.h]: Should this just be removed?

File: include/aws/crt/JsonObject.h [hunk 2]
```diff
@@ -33,12 +33,12 @@ namespace Aws
             /**
              * Constructs empty JSON DOM.
              */
-            JsonObject();
+            JsonObject(Allocator *a = DefaultAllocator());
```
Comment (justinboswell) [inline include/aws/crt/JsonObject.h]: Allocators are usually taken by reference, as they cannot be null. Why not do that for these objects?

File: include/aws/crt/Types.h [hunk 3]
```diff
@@ -76,8 +76,8 @@ namespace Aws
         AWS_CRT_CPP_API ByteCursor ByteCursorFromByteBuf(const ByteBuf &) noexcept;
         AWS_CRT_CPP_API ByteCursor ByteCursorFromArray(const uint8_t *array, size_t len) noexcept;
 
-        AWS_CRT_CPP_API Vector<uint8_t> Base64Decode(const String &decode);
-        AWS_CRT_CPP_API String Base64Encode(const Vector<uint8_t> &encode);
+        AWS_CRT_CPP_API Vector<uint8_t> Base64Decode(const String &decode, Allocator *a = DefaultAllocator());
```
Comment (justinboswell) [inline include/aws/crt/Types.h]: Allocator should be a reference.

File: include/aws/crt/UUID.h [hunk 4]
```diff
@@ -25,7 +25,7 @@ namespace Aws
         class UUID final
         {
           public:
-            UUID() noexcept;
+            UUID(Allocator *a = DefaultAllocator()) noexcept;
```
Comment (justinboswell) [inline include/aws/crt/UUID.h]: Allocator should be a reference

File: include/aws/iot/MqttClient.h (hunk 1 above)
Comment (bretambrose) [reply include/aws/iot/MqttClient.h]: Yeah, oops

File: include/aws/crt/JsonObject.h (hunk 2 above)
Comment (bretambrose) [reply include/aws/crt/JsonObject.h]: Mostly just following the existing pattern established in DefaultHostResolver, HttpConnectionManager, etc...  These are the c versions.

**Pull request closed without merging**